### PR TITLE
GLR Bison parser

### PR DESCRIPTION
**Pull request by dwightguth**
This contains a number of improvements to the K bison parser generator, including the ability to generate GLR parsers powerful enough to parse C.

### CODE REVIEW

File: k-distribution/tests/regression-new/parse-c/c18-syntax.k [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                | "for" "("     ";" Exp ";" Exp ")" Stmt
+                | "for" "("     ";"     ";" Exp ")" Stmt
+                | "for" "(" Exp ";"     ";"     ")" Stmt
+                | "for" "(" Exp ";" Exp ";"     ")" Stmt
+                | "for" "(" Exp ";" Exp ";" Exp ")" Stmt
+                | "for" "(" Exp ";"     ";" Exp ")" Stmt
+                | "for" "(" Decl     ";"     ")" Stmt
+                | "for" "(" Decl Exp ";"     ")" Stmt
+                | "for" "(" Decl Exp ";" Exp ")" Stmt
+                | "for" "(" Decl     ";" Exp ")" Stmt
```
Comment (radumereuta) [inline k-distribution/tests/regression-new/parse-c/c18-syntax.k:298]: The for doesn't look very nice. What happened?

File: kernel/src/main/java/org/kframework/parser/inner/kernel/Scanner.java [hunk 2]
```diff
@@ -201,8 +207,8 @@ private void writeAction(StringBuilder flex, TerminalLike key) {
     private void writeStandaloneAction(StringBuilder flex, TerminalLike key) {
         flex.append(" {\n" +
             "  int kind = ").append(tokens.get(key)._1()+1).append(";\n" +
-            "  *((char **)&yylval) = malloc(strlen(yytext) + 1);\n" +
-            "  strcpy(*((char **)&yylval), yytext);\n" +
+            "  *((char **)yylval  ) = malloc(strlen(yytext) + 1);\n" +
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/parser/inner/kernel/Scanner.java]: Why the extra space? :)

**Review by radumereuta — APPROVED** [review]
I had a quick look at the grammar and looks mostly fine.
I'm not that familiar with Bison, maybe Everet should also have a look before merging.

File: k-distribution/tests/regression-new/parse-c/c18-syntax.k (hunk 1 above)
Comment (dwightguth) [reply k-distribution/tests/regression-new/parse-c/c18-syntax.k:298]: Combinatory explosion. The reference grammar has three optional norterminals, so we generate 2^3=8 productions. The alternative would be to introduce a new nonterminal that has an empty production, but bison dislikes this immensely and you end up with a parser that splits a lot more often.

File: kernel/src/main/java/org/kframework/kompile/KompileOptions.java [hunk 3]
```diff
@@ -135,6 +137,9 @@ public boolean isKore() {
         @Parameter(names="--gen-bison-parser", description="Emit bison parser for the PGM configuration variable within the syntax module of your definition into the kompiled definition.")
         public boolean genBisonParser;
 
+        @Parameter(names="--gen-glr-bison-parser", description="Emit GLR bison parser for the PGM configuration variable within the syntax module of your definition into the kompiled definition.")
+        public boolean genGlrBisonParser;
```
Comment (ehildenb) [inline kernel/src/main/java/org/kframework/kompile/KompileOptions.java:141]: We should emit a parser for each `$` variable that shows up in the configuration, not just `$PGM`.

File: kernel/src/main/java/org/kframework/kore/convertors/KILtoKORE.java [hunk 4]
```diff
@@ -300,9 +303,15 @@ public void applyUserList(Set<org.kframework.definition.Sentence> res,
         org.kframework.definition.Production prod1, prod3;
 
         // Es ::= E "," Es
-        prod1 = Production(KLabel(p.getKLabel(kore), immutable(p.getParams())), sort,
-                Seq(NonTerminal(elementSort), Terminal(userList.getSeparator()), NonTerminal(sort)),
-                attrs.add("right"));
+        if (bisonLists) {
+          prod1 = Production(KLabel(p.getKLabel(kore), immutable(p.getParams())), sort,
+                  Seq(NonTerminal(sort), Terminal(userList.getSeparator()), NonTerminal(elementSort)),
+                  attrs.add("right"));
```
Comment (ehildenb) [inline kernel/src/main/java/org/kframework/kore/convertors/KILtoKORE.java]: Should this be `left` instead?

**Review by ehildenb — COMMENTED** [review]
We should generate parsers for _all_ `$..` vars in the initial configuration, not just `$PGM`, for both the LR and GLR parsers.

**Review by ehildenb — APPROVED** [review]

**Pull request merged**